[SPARK-19602][SQL][TESTS] Add tests for qualified column names

skambha opened the pull request:
## What changes were proposed in this pull request?
- Add tests covering different scenarios with qualified column names
- Please see Section 2 in the design doc for the various test scenarios [here](https://issues.apache.org/jira/secure/attachment/12854681/Design_ColResolution_JIRA19602.pdf)
- As part of SPARK-19602, changes are made to support three part column name. In order to aid in the review and to reduce the diff, the test scenarios are separated out into this PR.

## How was this patch tested?
- This is a **test only** change. The individual test suites were run successfully.

skambha committed:
Add tests to cover scenarios for the column resolution

skambha committed:
Add tests for global temp views

skambha committed:
Add test with local view

skambha committed:
newline in end

skambha committed:
uppercase struct

skambha committed:
fix style and use ctas

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

gatorsmile commented:
Some test cases are negative; some test cases are expected to be supported. Could you split them from the positive test cases? Now some test cases are pretty large.

SparkQA commented:
**[Test build #73479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73479/testReport)** for PR 17067 at commit [`2f9937e`](https://github.com/apache/spark/commit/2f9937e5dc0fc02218c644083432f1bff241409f).

gatorsmile commented:
Uh. Forgot one more point. 

Could you move these test cases to our new end-to-end SQL query suite into `SQLQueryTestSuite` ?

gatorsmile commented:
Thank you for working on this! I like the coverage of qualified column names.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73479/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #73494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73494/testReport)** for PR 17067 at commit [`2f9937e`](https://github.com/apache/spark/commit/2f9937e5dc0fc02218c644083432f1bff241409f).

SparkQA commented:
**[Test build #73494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73494/testReport)** for PR 17067 at commit [`2f9937e`](https://github.com/apache/spark/commit/2f9937e5dc0fc02218c644083432f1bff241409f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73494/
Test PASSed.

skambha committed:
Add column resolution tests to the SQLQueryTestSuite and also separate out the -ve cases from +ve

skambha committed:
Remove the -ve cases from the SQLQueryTestSuite because the exprId can change in exception

skambha committed:
Tests are moved to the SQLQueryTestSuite

skambha committed:
Remove empty lines

SparkQA commented:
**[Test build #73561 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73561/testReport)** for PR 17067 at commit [`e4f347e`](https://github.com/apache/spark/commit/e4f347e648efba81c1be1ff679d7dd88967d408b).

skambha committed:
Remove unused import

gatorsmile commented:
We can move the remaining test cases to `SQLQueryTestSuite `, right?

skambha commented:
Thanks much Xiao for the review and comments. 

I have made the following changes: 
- Separated out the -ve cases from the +ve cases. 
- Moved positive tests and also the cases that should be supported into the SQLQueryTestSuite framework.  A new test file columnresolution.sql and the corresponding master out file is added. 
- Clean up the ColumnResolutionSuite to remove cases that are covered in the SQLQueryTestSuite
- I have kept the -ve cases in the ColumnResolutionSuite because the exprId shows up in the exception.
- I also wanted to cover a case against a hive serde table so I have kept those tests in the ColumnResolutionSuite

Please advise if we should move any others.  Thanks.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala:
```diff
@@ -52,6 +52,19 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
+      checkAnswer(spark.sql("SELECT i1 FROM table1"), Row(2))
+      checkAnswer(spark.sql("SELECT table1.i1 FROM table1"), Row(2))
+      checkAnswer(spark.sql("SELECT a.i1 FROM table1 AS a"), Row(2))
+      checkAnswer(spark.sql("SELECT i1 FROM table1 AS a"), Row(2))
```
How about these test cases for temporary views? 

```
-- Test data.
CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES
(1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2), (null, 1), (3, null), (null, null)
AS testData(a, b);
```

skambha replied:
Sure, let me look at converting these too. Thanks.

gatorsmile replied:
Also doable for global temporary view, I think

skambha replied:
Hi Xiao,  I have moved my new local temp view tests and the global temp view tests to the SQLQueryTestSuite framework as well.  Please take a look. Thanks.

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ColumnResolutionSuite.scala:
```diff
@@ -0,0 +1,173 @@
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SQLTestUtils
+
+class ColumnResolutionSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
```
For the test cases you want to keep here, you can move it to `sql/core`. Why we need to test hive serde tables? Compared with data source tables, it is touching different code paths to resolve columns?

skambha replied:
The logic to resolve the column in the LogicalPlan is same - there is no change there.  I wanted to test the hive table to make sure that the qualifier information is correctly set. We update the qualifier info in MetastoreRelation so wanted to have coverage for hive table.

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql:
```diff
@@ -0,0 +1,82 @@
+-- Scenario: column resolution scenarios with datasource table
+CREATE DATABASE mydb1;
+use mydb1;
```
Please use upper case for SQL keywords.

SparkQA commented:
**[Test build #73561 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73561/testReport)** for PR 17067 at commit [`e4f347e`](https://github.com/apache/spark/commit/e4f347e648efba81c1be1ff679d7dd88967d408b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73561/
Test FAILed.

skambha committed:
Move view tests also to SQLQueryTestSuite

skambha committed:
Rebase with master

skambha committed:
upper casing SQL

SparkQA commented:
**[Test build #73622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73622/testReport)** for PR 17067 at commit [`1ae20ec`](https://github.com/apache/spark/commit/1ae20ec0a013358404f2b0d0e1ebadd60cdcd0b6).

SparkQA commented:
**[Test build #73622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73622/testReport)** for PR 17067 at commit [`1ae20ec`](https://github.com/apache/spark/commit/1ae20ec0a013358404f2b0d0e1ebadd60cdcd0b6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73622/
Test PASSed.

skambha committed:
Add -ve tests to the SQLQueryTestSuite framework

skambha committed:
This file isnt updated properly in Spark-19766

skambha committed:
Change to SQLQueryTestSuite to mask the id

skambha committed:
Delete the ColumnResolutionSuite

SparkQA commented:
**[Test build #73722 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73722/testReport)** for PR 17067 at commit [`5594eb0`](https://github.com/apache/spark/commit/5594eb0864376bbac617bf744755330f1e7bff49).

skambha commented:
- Changes to the SQLQueryTestSuite framework to mask the exprId so I can add the -ve cases as well using this framework.
- Added -ve test cases to the SQLQueryTestSuite framework and so removed the hive specific test suite.  For the hive table testcase, I will add that test as part of the actual code changes PR.
- I synced up the codeline and there was one test output inner-join.sql.out that needed a comment to be updated, so I have updated that as well.

SparkQA commented:
**[Test build #73722 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73722/testReport)** for PR 17067 at commit [`5594eb0`](https://github.com/apache/spark/commit/5594eb0864376bbac617bf744755330f1e7bff49).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73722/
Test PASSed.

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution-negative.sql:
```diff
@@ -0,0 +1,36 @@
+-- reset
+set spark.sql.crossJoin.enabled = false;
+DROP DATABASE mydb1 CASCADE;
+DROP DATABASE mydb2 CASCADE;
```
Please add one empty space after this line

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution-negative.sql:
```diff
@@ -0,0 +1,36 @@
+SELECT mydb1.t1.i1 FROM t1;
+
+-- reset
+set spark.sql.crossJoin.enabled = false;
```
Move this to line 24

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql:
```diff
@@ -0,0 +1,25 @@
+-- Tests for qualified column names for the view code-path
+-- Test scenario with Temporary view
+CREATE OR REPLACE TEMPORARY VIEW table1 AS SELECT 2 AS i1;
```
`table1 ` -> `view1`

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql:
```diff
@@ -0,0 +1,25 @@
+DROP VIEW table1;
+
+-- Test scenario with Global Temp view
+CREATE OR REPLACE GLOBAL TEMPORARY VIEW t1 as SELECT 1 as i1;
```
`t1` -> `view1`

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:
```diff
@@ -228,12 +228,13 @@ class SQLQueryTestSuite extends QueryTest with SharedSQLContext {
+        val msg = if (a.plan.nonEmpty) a.getSimpleMessage else a.getMessage
         (StructType(Seq.empty),
-          Seq(a.getClass.getName, a.getSimpleMessage.replaceAll("#\\d+", "#x")))
+          Seq(a.getClass.getName, msg.replaceAll("#\\d+", "#x")))
```
Nit: `(StructType(Seq.empty), Seq(a.getClass.getName, msg.replaceAll("#\\d+", "#x")))`

gatorsmile commented on sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out:
```diff
@@ -0,0 +1,240 @@
+struct<>
+-- !query 9 output
+org.apache.spark.sql.AnalysisException
+Reference 't1.i1' is ambiguous, could be: i1#x, i1#x.; line 1 pos 7
```
To the other reviewer, this is the reason why we need to make a change in `SQLQueryTestSuite.scala`

gatorsmile commented on sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql:
```diff
@@ -0,0 +1,84 @@
+INSERT INTO t5 VALUES(1, (2, 3));
+SELECT t5.i1 FROM t5;
+SELECT t5.t5.i1 FROM t5;
+SELECT t5.t5.i1 FROM mydb1.t5;
```
Add two cases for verifying `*`

gatorsmile commented:
Generally, it looks good to me.

skambha committed:
address comments

SparkQA commented:
**[Test build #73779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73779/testReport)** for PR 17067 at commit [`b2e411c`](https://github.com/apache/spark/commit/b2e411c859f4266428db6b4fbf55d89a101e50bf).

SparkQA commented:
**[Test build #73779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73779/testReport)** for PR 17067 at commit [`b2e411c`](https://github.com/apache/spark/commit/b2e411c859f4266428db6b4fbf55d89a101e50bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73779/
Test PASSed.

gatorsmile commented:
LGTM

gatorsmile commented:
Thanks! Merging to master.

skambha commented:
Thanks a lot Xiao.
